contracts/GenesisManager.sol

```diff
@@ -0,0 +1,268 @@
+        require(!tokenDistribution.isOver());
+
+        // Mint the initial supply
+        token.mint(this, initialSupply);
+
+        // Transfer the crowd supply to the token distribution contract
+        token.transfer(tokenDistribution, crowdSupply);
+
+        stage = Stages.GenesisStart;
+    }
+
+    /**
+     * @dev Add a team grant for tokens with a vesting schedule
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addTeamGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        external
+        onlyOwner
+        atStage(Stages.GenesisStart)
+    {
+        uint256 updatedGrantsAmount = teamGrantsAmount.add(_amount);
+        // Amount of tokens included in team grants cannot exceed the team supply during genesis
+        require(updatedGrantsAmount <= teamSupply);
+
+        teamGrantsAmount = updatedGrantsAmount;
+
+        addVestingGrant(_receiver, _amount, _timeToCliff, _vestingDuration);
+    }
+
+    /**
+     * @dev Add an investor grant for tokens with a vesting schedule
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addInvestorGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        external
+        onlyOwner
+        atStage(Stages.GenesisStart)
+    {
+        uint256 updatedGrantsAmount = investorsGrantsAmount.add(_amount);
+        // Amount of tokens included in investor grants cannot exceed the investor supply during genesis
+        require(updatedGrantsAmount <= investorsSupply);
+
+        investorsGrantsAmount = updatedGrantsAmount;
+
+        addVestingGrant(_receiver, _amount, _timeToCliff, _vestingDuration);
+    }
+
+    /**
+     * @dev Add a grant for tokens with a vesting schedule. An internal helper function used by addTeamGrant and addInvestorGrant
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addVestingGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        internal
+    {
+        // Receiver must not have already received a grant with a vesting schedule
+        require(vestingHolders[_receiver] == address(0));
```

Comment:
we'll need two different addresses if the same person is receiving multiple grants. that's ok.